[MSE] `ASSERT(canSafelyBeUsed());` ASSERTION when using webm with MSE in a worker

jyavenard opened the pull request:
#### de9e3d6bc4880dab143d3f4e3643adf588078dd9
<pre>
[MSE] `ASSERT(canSafelyBeUsed());` ASSERTION when using webm with MSE in a worker
<a href="https://bugs.webkit.org/show_bug.cgi?id=269638">https://bugs.webkit.org/show_bug.cgi?id=269638</a>
<a href="https://rdar.apple.com/123134428">rdar://123134428</a>

Reviewed by Eric Carlson.

With a MediaSource running in a DedicatedWorker, a TrackPrivate may be referenced by two Tracks:
- The canonical track living in the SourceBuffer
- Its mirror that is sent to the HTMLMediaElement on the main thread.

The interface between the Track and the TrackPrivate is done via the TrackPrivateBaseClient
which inherit from CanMakeWeakPtr. A WeakPtr was used to store the TrackPrivateClient
in the TrackPrivate.
As the TrackPrivate could only track one client, it was the last caller to TrackPrivate::setClient
that won. So you could end up having the TrackPrivate attempting to use the WeakPtr created on the
worker thread while the TrackPrivate is being used in the main thread.

We add support for multiple clients, and to ensure that the client will only ever be called
on the right thread, TrackPrivate::addClient now takes a dispatcher that will ensure the
task to run and the WeakPtr will always be accessed on the right thread.
TrackPrivate::addClient|removeClient are thread safe.

Previously, {Audio|Video|InbandTrack}PrivateTrack each managed their respective client.
We move the logic to the base TrackPrivateBase class to remove mostly duplicated code.

Move semantics with the client&apos;s callback is no longer possible, this is fine for
Audio and Video tracks as their message is only to enable/select them; however for
InbandTextTrack it becomes problematic as the WebVTT parser API and Cues all require
move semantics.
However, InbandTextTracks aren&apos;t currently supported with our MediaSource implementation
(no User-Agent does), so there will only ever be one client to those tracks, and the
client will be running in the main thread.
We add assertions for methods were moved semantics is required, and to prevent unexpected
failures, only the first client will be called, and only if on the main thread.

The Remote{Audio|Video|InbandText}PrivateTrack which are the only tracks used when MSE in
a worker is active, were using AtomString which can&apos;t be transferred across thread.
We change the type of their strings from AtomString to String, and ensure that
the string copied are isolated copies.

For non-MSE, no change in observable behaviour.
A test will be added in a follow-up change: <a href="https://webkit.org/b/269564)">https://webkit.org/b/269564)</a>

* Source/WebCore/html/track/AudioTrack.cpp:
(WebCore::AudioTrack::AudioTrack):
(WebCore::AudioTrack::~AudioTrack):
(WebCore::AudioTrack::setPrivate):
* Source/WebCore/html/track/InbandTextTrack.cpp:
(WebCore::InbandTextTrack::InbandTextTrack):
(WebCore::InbandTextTrack::~InbandTextTrack):
(WebCore::InbandTextTrack::setPrivate):
* Source/WebCore/html/track/TrackBase.cpp:
(WebCore::TrackBase::addClientToTrackPrivateBase):
(WebCore::TrackBase::removeClientFromTrackPrivateBase):
* Source/WebCore/html/track/TrackBase.h:
* Source/WebCore/html/track/VideoTrack.cpp:
(WebCore::VideoTrack::VideoTrack):
(WebCore::VideoTrack::~VideoTrack):
(WebCore::VideoTrack::setPrivate):
* Source/WebCore/platform/graphics/AudioTrackPrivate.h:
(WebCore::AudioTrackPrivate::setEnabled):
(WebCore::AudioTrackPrivate::setConfiguration):
(WebCore::AudioTrackPrivate::setClient): Deleted.
(WebCore::AudioTrackPrivate::clearClient): Deleted.
* Source/WebCore/platform/graphics/AudioTrackPrivateClient.h:
(isType):
* Source/WebCore/platform/graphics/InbandTextTrackPrivate.h:
(WebCore::InbandTextTrackPrivate::setClient): Deleted.
(WebCore::InbandTextTrackPrivate::clearClient): Deleted.
* Source/WebCore/platform/graphics/InbandTextTrackPrivateClient.h:
(isType):
* Source/WebCore/platform/graphics/TrackPrivateBase.cpp:
(WebCore::TrackPrivateBase::notifyClients):
(WebCore::TrackPrivateBase::notifyMainThreadClient):
(WebCore::TrackPrivateBase::addClient):
(WebCore::TrackPrivateBase::removeClient):
(WebCore::TrackPrivateBase::hasClients const):
(WebCore::TrackPrivateBase::hasOneClient const):
* Source/WebCore/platform/graphics/TrackPrivateBase.h:
* Source/WebCore/platform/graphics/TrackPrivateBaseClient.h:
* Source/WebCore/platform/graphics/VideoTrackPrivate.h:
(WebCore::VideoTrackPrivate::setSelected):
(WebCore::VideoTrackPrivate::setConfiguration):
(WebCore::VideoTrackPrivate::setClient): Deleted.
(WebCore::VideoTrackPrivate::clearClient): Deleted.
* Source/WebCore/platform/graphics/VideoTrackPrivateClient.h:
(isType):
* Source/WebCore/platform/graphics/avfoundation/InbandMetadataTextTrackPrivateAVF.cpp:
(WebCore::InbandMetadataTextTrackPrivateAVF::addDataCue):
(WebCore::InbandMetadataTextTrackPrivateAVF::updatePendingCueEndTimes):
(WebCore::InbandMetadataTextTrackPrivateAVF::flushPartialCues):
* Source/WebCore/platform/graphics/avfoundation/InbandTextTrackPrivateAVF.cpp:
(WebCore::InbandTextTrackPrivateAVF::processCue):
(WebCore::InbandTextTrackPrivateAVF::processAttributedStrings):
(WebCore::InbandTextTrackPrivateAVF::removeCompletedCues):
(WebCore::InbandTextTrackPrivateAVF::resetCueValues):
(WebCore::InbandTextTrackPrivateAVF::processNativeSamples):
* Source/WebCore/platform/graphics/avfoundation/objc/InbandChapterTrackPrivateAVFObjC.mm:
(WebCore::InbandChapterTrackPrivateAVFObjC::processChapters):
* Source/WebCore/platform/graphics/gstreamer/AudioTrackPrivateGStreamer.cpp:
(WebCore::AudioTrackPrivateGStreamer::updateConfigurationFromTags):
* Source/WebCore/platform/graphics/gstreamer/InbandTextTrackPrivateGStreamer.cpp:
(WebCore::InbandTextTrackPrivateGStreamer::tagsChanged):
(WebCore::InbandTextTrackPrivateGStreamer::notifyTrackOfSample):
* Source/WebCore/platform/graphics/gstreamer/VideoTrackPrivateGStreamer.cpp:
(WebCore::VideoTrackPrivateGStreamer::updateConfigurationFromTags):
* Source/WebKit/GPUProcess/media/RemoteAudioTrackProxy.cpp:
(WebKit::RemoteAudioTrackProxy::RemoteAudioTrackProxy):
(WebKit::RemoteAudioTrackProxy::~RemoteAudioTrackProxy):
* Source/WebKit/GPUProcess/media/RemoteAudioTrackProxy.h:
* Source/WebKit/GPUProcess/media/RemoteTextTrackProxy.cpp:
(WebKit::RemoteTextTrackProxy::RemoteTextTrackProxy):
(WebKit::RemoteTextTrackProxy::~RemoteTextTrackProxy):
* Source/WebKit/GPUProcess/media/RemoteTextTrackProxy.h:
* Source/WebKit/GPUProcess/media/RemoteVideoTrackProxy.cpp:
(WebKit::RemoteVideoTrackProxy::RemoteVideoTrackProxy):
(WebKit::RemoteVideoTrackProxy::~RemoteVideoTrackProxy):
* Source/WebKit/GPUProcess/media/RemoteVideoTrackProxy.h:
* Source/WebKit/GPUProcess/media/TextTrackPrivateRemoteConfiguration.h:
* Source/WebKit/GPUProcess/media/TextTrackPrivateRemoteConfiguration.serialization.in: Change data type to AtomString
* Source/WebKit/GPUProcess/media/TrackPrivateRemoteConfiguration.h:
* Source/WebKit/GPUProcess/media/TrackPrivateRemoteConfiguration.serialization.in: Change data type to AtomString
* Source/WebKit/WebProcess/GPU/media/AudioTrackPrivateRemote.cpp:
(WebKit::AudioTrackPrivateRemote::updateConfiguration):
* Source/WebKit/WebProcess/GPU/media/AudioTrackPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/TextTrackPrivateRemote.cpp:
(WebKit::TextTrackPrivateRemote::updateConfiguration):
(WebKit::TextTrackPrivateRemote::addGenericCue):
(WebKit::TextTrackPrivateRemote::updateGenericCue):
(WebKit::TextTrackPrivateRemote::removeGenericCue):
(WebKit::TextTrackPrivateRemote::parseWebVTTFileHeader):
(WebKit::TextTrackPrivateRemote::parseWebVTTCueData):
(WebKit::TextTrackPrivateRemote::parseWebVTTCueDataStruct):
(WebKit::TextTrackPrivateRemote::addDataCue):
(WebKit::TextTrackPrivateRemote::addDataCueWithType):
(WebKit::TextTrackPrivateRemote::updateDataCue):
(WebKit::TextTrackPrivateRemote::removeDataCue):
* Source/WebKit/WebProcess/GPU/media/TextTrackPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/VideoTrackPrivateRemote.cpp:
(WebKit::VideoTrackPrivateRemote::updateConfiguration):
* Source/WebKit/WebProcess/GPU/media/VideoTrackPrivateRemote.h:

Canonical link: <a href="https://commits.webkit.org/274956@main">https://commits.webkit.org/274956@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/be3ba6d538d0d3f425a3c27ad1f8bbd40c18e3fa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40460 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19472 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42838 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43010 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36549 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42767 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22427 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16801 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33568 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41034 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16423 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34879 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14152 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14218 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35851 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44286 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36684 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36174 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39921 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15275 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12514 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38222 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16894 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9074 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16944 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16538 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/html/track/AudioTrack.cpp:
```diff
@@ -68,25 +68,27 @@ AudioTrack::AudioTrack(ScriptExecutionContext* context, AudioTrackPrivate& track
     , m_configuration(AudioTrackConfiguration::create())
 {
-    m_private->setClient(*this);
+    addAsClient(*this, trackPrivate);
```
Could we use `registerClient` / `unregisterClient` naming like we usually do in the code base?

cdumez replied:
or `addClient()` / `removeClient()`

jyavenard replied:
actually, addClient() is already a method of `AudioTrack`; the aim here is to add ourselves as client to the trackPrivate

cdumez commented on Source/WebCore/html/track/AudioTrack.cpp:
```diff
@@ -68,25 +68,27 @@ AudioTrack::AudioTrack(ScriptExecutionContext* context, AudioTrackPrivate& track
 AudioTrack::~AudioTrack()
 {
-    m_private->clearClient();
+    removeAsClient(m_private);
```
`Ref { m_private }` per smart pointer guidelines.

cdumez replied:
You can also introduce a `protectedPrivate()` getter which returns a `Ref`

jyavenard replied:
the name is a tad confusing, as we remove ourselves as client of m_private ; m_private isn't our client, other way round.

That's why I had used `addAsClient` ; can't think of a better name though

cdumez commented on Source/WebCore/html/track/AudioTrack.cpp:
```diff
@@ -68,25 +68,27 @@ AudioTrack::AudioTrack(ScriptExecutionContext* context, AudioTrackPrivate& track
         return;
 
-    m_private->clearClient();
+    removeAsClient(m_private);
```
`Ref { m_private }`

cdumez commented on Source/WebCore/html/track/InbandTextTrack.cpp:
```diff
@@ -60,23 +60,23 @@ InbandTextTrack::InbandTextTrack(ScriptExecutionContext& context, InbandTextTrac
 InbandTextTrack::~InbandTextTrack()
 {
-    m_private->clearClient();
+    removeAsClient(m_private);
```
`Ref { m_private }`

cdumez commented on Source/WebCore/html/track/TrackBase.cpp:
```diff
@@ -182,6 +184,20 @@ WTFLogChannel& TrackBase::logChannel() const
+void TrackBase::addAsClient(TrackPrivateBaseClient& client, TrackPrivateBase& track)
+{
+    if (auto context = scriptExecutionContext()) {
+        m_clientId = track.addClient([context = context->identifier()](auto&& task) {
```
Would be good to call this `contextIdentifier` or `contextID` rather than simply `context`, which normally represents a ScriptExecutionContext object.

cdumez commented on Source/WebCore/html/track/TrackBase.h:
```diff
@@ -107,6 +112,7 @@ class TrackBase
     const void* m_logIdentifier { nullptr };
 #endif
     WeakPtr<TrackListBase, WeakPtrImplWithEventTargetData> m_trackList;
+    size_t m_clientId;
```
Any reason not to use a uint64_t? Or even better, an ObjectIdentifier?

cdumez replied:
It looks like this is left uninitialized until `addAsClient()` gets called? If so, this isn't good.

cdumez commented on Source/WebCore/html/track/VideoTrack.cpp:
```diff
@@ -60,24 +60,24 @@ VideoTrack::VideoTrack(ScriptExecutionContext* context, VideoTrackPrivate& track
 VideoTrack::~VideoTrack()
 {
-    m_private->clearClient();
+    removeAsClient(m_private);
```
`Ref { m_private }`

cdumez commented on Source/WebCore/html/track/VideoTrack.cpp:
```diff
@@ -60,24 +60,24 @@ VideoTrack::VideoTrack(ScriptExecutionContext* context, VideoTrackPrivate& track
         return;
 
-    m_private->clearClient();
+    removeAsClient(m_private);
```
`Ref { m_private }`

cdumez commented on Source/WebCore/platform/graphics/AudioTrackPrivate.h:
```diff
@@ -38,17 +38,15 @@ struct AudioInfo;
-        if (m_client)
-            m_client->enabledChanged(enabled);
+        notifyClients([enabled](auto& client) {
+            ASSERT(is<AudioTrackPrivateClient>(client));
```
Not needed, downcast already does this for you. That's the whole point of downcast vs static_cast.

cdumez commented on Source/WebCore/platform/graphics/AudioTrackPrivate.h:
```diff
@@ -38,17 +38,15 @@ struct AudioInfo;
-            m_client->enabledChanged(enabled);
+        notifyClients([enabled](auto& client) {
+            ASSERT(is<AudioTrackPrivateClient>(client));
+            downcast<AudioTrackPrivateClient>(client).enabledChanged(enabled);
```
`enabledChanged` sounds a little weird to me. Maybe `enabledStateChanged`?

cdumez commented on Source/WebCore/platform/graphics/AudioTrackPrivate.h:
```diff
@@ -69,8 +67,10 @@ class AudioTrackPrivate : public TrackPrivateBase {
-        if (m_client)
-            m_client->configurationChanged(m_configuration);
+        notifyClients([configuration = m_configuration](auto& client) {
+            ASSERT(is<AudioTrackPrivateClient>(client));
```
Not needed.

cdumez commented on Source/WebCore/platform/graphics/AudioTrackPrivateClient.h:
```diff
@@ -36,10 +36,15 @@ struct PlatformAudioTrackConfiguration;
 
 class AudioTrackPrivateClient : public TrackPrivateBaseClient {
 public:
+    constexpr Type type() const final { return Type::Audio; }
```
I assume `type()` can be private here? Does constexpr bring us anything here? I'm curious.

cdumez commented on Source/WebCore/platform/graphics/InbandTextTrackPrivateClient.h:
```diff
@@ -46,6 +46,8 @@ class InbandTextTrackPrivateClient : public TrackPrivateBaseClient {
 public:
     virtual ~InbandTextTrackPrivateClient() = default;
 
+    constexpr Type type() const final { return Type::Text; }
```
I assume this can be private. Same question about constexpr.

cdumez commented on Source/WebCore/platform/graphics/TrackPrivateBase.cpp:
```diff
@@ -53,6 +54,76 @@ bool TrackPrivateBase::operator==(const TrackPrivateBase& track) const
+    Locker locker { m_lock };
+    if (m_clients.size() > index)
+        return;
+    m_clients[index] = std::make_tuple<RefPtr<SharedDispatcher>, WeakPtr<TrackPrivateBaseClient>, bool>({ }, { }, false);
```
Wouldn't `= { }` work?

cdumez commented on Source/WebCore/platform/graphics/TrackPrivateBase.cpp:
```diff
@@ -53,6 +54,76 @@ bool TrackPrivateBase::operator==(const TrackPrivateBase& track) const
+    return m_clients.size();
+}
+
+bool TrackPrivateBase::hasOneClient() const
```
These kind of functions: hasClients() / hasOneClient() when locks / multithreading are involved seem a little fishy? As soon as the function returns, the answer may already be wrong (since the lock is held internally and released on return).

cdumez commented on Source/WebCore/platform/graphics/TrackPrivateBase.h:
```diff
@@ -50,7 +53,8 @@ class WEBCORE_EXPORT TrackPrivateBase
 
-    virtual TrackPrivateBaseClient* client() const = 0;
+    size_t addClient(TrackPrivateBaseClient::Dispatcher&&, TrackPrivateBaseClient&);
+    void removeClient(uint32_t); // Can be called multiple times with the same id.
```
Would really be nice to use ObjectIdentifier for identifiers. Also, earlier the id was a size_t and here it's an uint32_t? Would be nice for the PR to be consistent.

cdumez commented on Source/WebCore/platform/graphics/TrackPrivateBase.h:
```diff
@@ -80,9 +85,34 @@ class WEBCORE_EXPORT TrackPrivateBase
     TrackPrivateBase() = default;
 
+    template <typename T>
+    class Shared final : public ThreadSafeRefCounted<Shared<T>> {
```
Did you just re-invent WTF::Box? At least it looks super similar so it's worth double checking.

cdumez commented on Source/WebCore/platform/graphics/TrackPrivateBase.h:
```diff
@@ -80,9 +85,34 @@ class WEBCORE_EXPORT TrackPrivateBase
+
+    bool hasClients() const;
+    bool hasOneClient() const;
+    mutable Lock m_lock;
```
Please separate member functions and data members with a blank line.

cdumez commented on Source/WebCore/platform/graphics/TrackPrivateBaseClient.h:
```diff
@@ -36,7 +36,12 @@ using TrackID = uint64_t;
+    using Dispatcher = Function<void(Task&&)>;
+
     virtual ~TrackPrivateBaseClient() = default;
+    enum Type { Text, Audio, Video };
```
Please only use enum classes in new code, and specify their underlying type (here uint8_t).

cdumez commented on Source/WebCore/platform/graphics/VideoTrackPrivate.h:
```diff
@@ -38,17 +38,15 @@ struct VideoInfo;
-        if (m_client)
-            m_client->selectedChanged(m_selected);
+        notifyClients([selected](auto& client) {
+            ASSERT(is<VideoTrackPrivateClient>(client));
```
Please remove.

cdumez commented on Source/WebCore/platform/graphics/VideoTrackPrivate.h:
```diff
@@ -70,8 +68,10 @@ class VideoTrackPrivate : public TrackPrivateBase {
-        if (m_client)
-            m_client->configurationChanged(m_configuration);
+        notifyClients([configuration = m_configuration](auto& client) {
+            ASSERT(is<VideoTrackPrivateClient>(client));
```
Please remove.

cdumez commented on Source/WebCore/platform/graphics/VideoTrackPrivateClient.h:
```diff
@@ -35,10 +35,15 @@ struct PlatformVideoTrackConfiguration;
 
 class VideoTrackPrivateClient : public TrackPrivateBaseClient {
 public:
+    constexpr Type type() const final { return Type::Video; }
```
private?

cdumez commented on Source/WebCore/platform/graphics/avfoundation/InbandMetadataTextTrackPrivateAVF.cpp:
```diff
@@ -56,27 +56,35 @@ InbandMetadataTextTrackPrivateAVF::~InbandMetadataTextTrackPrivateAVF() = defaul
-    if (!client())
+    if (!hasClients())
         return;
+    ASSERT(hasOneClient());
```
Could a client be added on another thread in between the `hasClients()` check and this assertion?

cdumez commented on Source/WebCore/platform/graphics/avfoundation/InbandMetadataTextTrackPrivateAVF.cpp:
```diff
@@ -56,27 +56,35 @@ InbandMetadataTextTrackPrivateAVF::~InbandMetadataTextTrackPrivateAVF() = defaul
     if (time >= m_currentCueStartTime) {
-        if (client()) {
+        if (hasClients()) {
+            ASSERT(hasOneClient());
```
Same question here about a potential race since you're not holding a lock.

cdumez commented on Source/WebCore/platform/graphics/avfoundation/InbandMetadataTextTrackPrivateAVF.cpp:
```diff
@@ -90,12 +98,17 @@ void InbandMetadataTextTrackPrivateAVF::updatePendingCueEndTimes(const MediaTime
-        for (auto& partialCue : m_incompleteCues)
-            client()->removeDataCue(partialCue.startTime, MediaTime::positiveInfiniteTime(), *partialCue.cueData);
+    if (hasClients()) {
+        ASSERT(hasOneClient());
```
ditto.

cdumez commented on Source/WebCore/platform/graphics/avfoundation/objc/InbandChapterTrackPrivateAVFObjC.mm:
```diff
@@ -50,21 +50,25 @@
         INFO_LOG(identifier, "created cue ", cueData);
-        client()->parseWebVTTCueData(WTFMove(cueData));
+        notifyMainThreadClient([cueData = WTFMove(cueData)](TrackPrivateBaseClient& client) mutable {
+            ASSERT(is<InbandTextTrackPrivateClient>(client));
```
Please remove.

cdumez commented on Source/WebCore/platform/graphics/gstreamer/InbandTextTrackPrivateGStreamer.cpp:
```diff
@@ -122,7 +124,11 @@ void InbandTextTrackPrivateGStreamer::notifyTrackOfSample()
+        ASSERT(isMainThread());
+        ASSERT(!hasClients() || hasOneClient());
+        notifyMainThreadClient([&](auto& client) {
+            downcast<InbandTextTrackPrivateClient>(client).parseWebVTTCueData(mappedBuffer.data(), mappedBuffer.size());
```
Could we pass a std::span? We try to avoid pointers + size in new code.

cdumez commented on Source/WebKit/GPUProcess/media/RemoteAudioTrackProxy.h:
```diff
@@ -85,6 +85,7 @@ class RemoteAudioTrackProxy final
     WebCore::TrackID m_id;
     WebCore::MediaPlayerIdentifier m_mediaPlayerIdentifier;
     bool m_enabled { false };
+    size_t m_clientId { 0 };
```
Same comments as earlier about id type.

cdumez commented on Source/WebKit/GPUProcess/media/RemoteTextTrackProxy.h:
```diff
@@ -93,6 +93,7 @@ class RemoteTextTrackProxy final
     Ref<WebCore::InbandTextTrackPrivate> m_trackPrivate;
     WebCore::TrackID m_id;
     WebCore::MediaPlayerIdentifier m_mediaPlayerIdentifier;
+    size_t m_clientId { 0 };
```
ditto.

jyavenard replied:
that's different though, this is the registrationID we received when we added ourselves to the TrackPrivateBase as client, and is needed to be removed. I'll change it to m_clientRegistrationId then

cdumez commented on Source/WebKit/GPUProcess/media/RemoteVideoTrackProxy.h:
```diff
@@ -85,6 +85,7 @@ class RemoteVideoTrackProxy final
     WebCore::TrackID m_id;
     WebCore::MediaPlayerIdentifier m_mediaPlayerIdentifier;
     bool m_selected { false };
+    size_t m_clientId { 0 };
```
ditto.

cdumez commented on Source/WebKit/WebProcess/GPU/media/AudioTrackPrivateRemote.cpp:
```diff
@@ -60,22 +61,29 @@ void AudioTrackPrivateRemote::updateConfiguration(AudioTrackPrivateRemoteConfigu
         m_id = configuration.trackId;
-        if (client())
-            client()->idChanged(m_id);
+        notifyClients([id = m_id](auto& client) {
```
We don't need to crossThreadCopy() m_id here? I didn't check the type. I just thought it was worth asking because you're doing it for m_id later in this PR.

jyavenard replied:
it's an int

cdumez commented on Source/WebKit/WebProcess/GPU/media/TextTrackPrivateRemote.cpp:
```diff
@@ -62,22 +63,29 @@ void TextTrackPrivateRemote::updateConfiguration(TextTrackPrivateRemoteConfigura
         m_id = configuration.trackId;
-        if (client())
-            client()->idChanged(m_id);
+        notifyClients([id = crossThreadCopy(m_id)](auto& client) {
```
See how you're cross-copying here. Might be fine but the mismatch made me worry.

jyavenard replied:
ok, it's a mistake to call callThreadCopy; copy/paste

cdumez commented on Source/WebKit/WebProcess/GPU/media/VideoTrackPrivateRemote.cpp:
```diff
@@ -60,22 +61,29 @@ void VideoTrackPrivateRemote::updateConfiguration(VideoTrackPrivateRemoteConfigu
         m_id = configuration.trackId;
-        if (client())
-            client()->idChanged(m_id);
+        notifyClients([id = m_id](auto& client) {
```
No need to crossThreadCopy here?

jyavenard replied:
no, it's an in

cdumez commented on Source/WebKit/WebProcess/GPU/media/AudioTrackPrivateRemote.h:
```diff
@@ -53,17 +53,17 @@ class AudioTrackPrivateRemote final : public WebCore::AudioTrackPrivate {
     AudioTrackKind kind() const final { return m_kind; }
-    AtomString label() const final { return m_label; }
-    AtomString language() const final { return m_language; }
+    AtomString label() const final { return AtomString { m_label }; }
```
Is there a benefit to returning AtomString here instead of a String? Same question applies elsewhere. Given that the data member is now a String, I'm a little worried about perf if we just keep atomizing the string every time we call the getter. 2 benefits from using an AtomString are:
1. Memory saving thanks to deduplication but that's really only useful for storage. Here the data member is a String.
2. Faster comparison between AtomStrings than between Strings.

I am not familiar with this code. Just thought I would make sure you thought of the benefits / drawbacks.

jyavenard replied:
No, there's no benefit to return an AtomString here; but every other implementations are using AtomStrings , including the DOM API.

Changing the base class definition is going to open a pandora box; and there's *lots* of stuff using AtomString

jyavenard replied:
Also, the string we originally got from the GPU process was a String, made into an AtomString before reaching us. 

In any case, that string needs to be sent to both the HTMLMediaElement's Track, and the SourceBuffer's Track so a copy would have been made one way or another along the way.

Ideally, we would have thread-safe ref-counted string.

eric-carlson reviewed (approved):
r=me on the media-specific parts, given you have addresses Chris’ feedback

jyavenard committed:
[MSE] `ASSERT(canSafelyBeUsed());` ASSERTION when using webm with MSE in a worker
https://bugs.webkit.org/show_bug.cgi?id=269638
rdar://123134428

Reviewed by Eric Carlson.

With a MediaSource running in a DedicatedWorker, a TrackPrivate may be referenced by two Tracks:
- The canonical track living in the SourceBuffer
- Its mirror that is sent to the HTMLMediaElement on the main thread.

The interface between the Track and the TrackPrivate is done via the TrackPrivateBaseClient
which inherit from CanMakeWeakPtr. A WeakPtr was used to store the TrackPrivateClient
in the TrackPrivate.
As the TrackPrivate could only track one client, it was the last caller to TrackPrivate::setClient
that won. So you could end up having the TrackPrivate attempting to use the WeakPtr created on the
worker thread while the TrackPrivate is being used in the main thread.

We add support for multiple clients, and to ensure that the client will only ever be called
on the right thread, TrackPrivate::addClient now takes a dispatcher that will ensure the
task to run and the WeakPtr will always be accessed on the right thread.
TrackPrivate::addClient|removeClient are thread safe.

Previously, {Audio|Video|InbandTrack}PrivateTrack each managed their respective client.
We move the logic to the base TrackPrivateBase class to remove mostly duplicated code.

Move semantics with the client's callback is no longer possible, this is fine for
Audio and Video tracks as their message is only to enable/select them; however for
InbandTextTrack it becomes problematic as the WebVTT parser API and Cues all require
move semantics.
However, InbandTextTracks aren't currently supported with our MediaSource implementation
(no User-Agent does), so there will only ever be one client to those tracks, and the
client will be running in the main thread.
We add assertions for methods were moved semantics is required, and to prevent unexpected
failures, only the first client will be called, and only if on the main thread.

The Remote{Audio|Video|InbandText}PrivateTrack which are the only tracks used when MSE in
a worker is active, were using AtomString which can't be transferred across thread.
We change the type of their strings from AtomString to String, and ensure that
the string copied are isolated copies.

For non-MSE, no change in observable behaviour.
A test will be added in a follow-up change: https://webkit.org/b/269564)

* Source/WebCore/html/track/AudioTrack.cpp:
(WebCore::AudioTrack::AudioTrack):
(WebCore::AudioTrack::~AudioTrack):
(WebCore::AudioTrack::setPrivate):
* Source/WebCore/html/track/InbandTextTrack.cpp:
(WebCore::InbandTextTrack::InbandTextTrack):
(WebCore::InbandTextTrack::~InbandTextTrack):
(WebCore::InbandTextTrack::setPrivate):
* Source/WebCore/html/track/TrackBase.cpp:
(WebCore::TrackBase::addClientToTrackPrivateBase):
(WebCore::TrackBase::removeClientFromTrackPrivateBase):
* Source/WebCore/html/track/TrackBase.h:
* Source/WebCore/html/track/VideoTrack.cpp:
(WebCore::VideoTrack::VideoTrack):
(WebCore::VideoTrack::~VideoTrack):
(WebCore::VideoTrack::setPrivate):
* Source/WebCore/platform/graphics/AudioTrackPrivate.h:
(WebCore::AudioTrackPrivate::setEnabled):
(WebCore::AudioTrackPrivate::setConfiguration):
(WebCore::AudioTrackPrivate::setClient): Deleted.
(WebCore::AudioTrackPrivate::clearClient): Deleted.
* Source/WebCore/platform/graphics/AudioTrackPrivateClient.h:
(isType):
* Source/WebCore/platform/graphics/InbandTextTrackPrivate.h:
(WebCore::InbandTextTrackPrivate::setClient): Deleted.
(WebCore::InbandTextTrackPrivate::clearClient): Deleted.
* Source/WebCore/platform/graphics/InbandTextTrackPrivateClient.h:
(isType):
* Source/WebCore/platform/graphics/TrackPrivateBase.cpp:
(WebCore::TrackPrivateBase::notifyClients):
(WebCore::TrackPrivateBase::notifyMainThreadClient):
(WebCore::TrackPrivateBase::addClient):
(WebCore::TrackPrivateBase::removeClient):
(WebCore::TrackPrivateBase::hasClients const):
(WebCore::TrackPrivateBase::hasOneClient const):
* Source/WebCore/platform/graphics/TrackPrivateBase.h:
* Source/WebCore/platform/graphics/TrackPrivateBaseClient.h:
* Source/WebCore/platform/graphics/VideoTrackPrivate.h:
(WebCore::VideoTrackPrivate::setSelected):
(WebCore::VideoTrackPrivate::setConfiguration):
(WebCore::VideoTrackPrivate::setClient): Deleted.
(WebCore::VideoTrackPrivate::clearClient): Deleted.
* Source/WebCore/platform/graphics/VideoTrackPrivateClient.h:
(isType):
* Source/WebCore/platform/graphics/avfoundation/InbandMetadataTextTrackPrivateAVF.cpp:
(WebCore::InbandMetadataTextTrackPrivateAVF::addDataCue):
(WebCore::InbandMetadataTextTrackPrivateAVF::updatePendingCueEndTimes):
(WebCore::InbandMetadataTextTrackPrivateAVF::flushPartialCues):
* Source/WebCore/platform/graphics/avfoundation/InbandTextTrackPrivateAVF.cpp:
(WebCore::InbandTextTrackPrivateAVF::processCue):
(WebCore::InbandTextTrackPrivateAVF::processAttributedStrings):
(WebCore::InbandTextTrackPrivateAVF::removeCompletedCues):
(WebCore::InbandTextTrackPrivateAVF::resetCueValues):
(WebCore::InbandTextTrackPrivateAVF::processNativeSamples):
* Source/WebCore/platform/graphics/avfoundation/objc/InbandChapterTrackPrivateAVFObjC.mm:
(WebCore::InbandChapterTrackPrivateAVFObjC::processChapters):
* Source/WebCore/platform/graphics/gstreamer/AudioTrackPrivateGStreamer.cpp:
(WebCore::AudioTrackPrivateGStreamer::updateConfigurationFromTags):
* Source/WebCore/platform/graphics/gstreamer/InbandTextTrackPrivateGStreamer.cpp:
(WebCore::InbandTextTrackPrivateGStreamer::tagsChanged):
(WebCore::InbandTextTrackPrivateGStreamer::notifyTrackOfSample):
* Source/WebCore/platform/graphics/gstreamer/VideoTrackPrivateGStreamer.cpp:
(WebCore::VideoTrackPrivateGStreamer::updateConfigurationFromTags):
* Source/WebKit/GPUProcess/media/RemoteAudioTrackProxy.cpp:
(WebKit::RemoteAudioTrackProxy::RemoteAudioTrackProxy):
(WebKit::RemoteAudioTrackProxy::~RemoteAudioTrackProxy):
* Source/WebKit/GPUProcess/media/RemoteAudioTrackProxy.h:
* Source/WebKit/GPUProcess/media/RemoteTextTrackProxy.cpp:
(WebKit::RemoteTextTrackProxy::RemoteTextTrackProxy):
(WebKit::RemoteTextTrackProxy::~RemoteTextTrackProxy):
* Source/WebKit/GPUProcess/media/RemoteTextTrackProxy.h:
* Source/WebKit/GPUProcess/media/RemoteVideoTrackProxy.cpp:
(WebKit::RemoteVideoTrackProxy::RemoteVideoTrackProxy):
(WebKit::RemoteVideoTrackProxy::~RemoteVideoTrackProxy):
* Source/WebKit/GPUProcess/media/RemoteVideoTrackProxy.h:
* Source/WebKit/GPUProcess/media/TextTrackPrivateRemoteConfiguration.h:
* Source/WebKit/GPUProcess/media/TextTrackPrivateRemoteConfiguration.serialization.in: Change data type to AtomString
* Source/WebKit/GPUProcess/media/TrackPrivateRemoteConfiguration.h:
* Source/WebKit/GPUProcess/media/TrackPrivateRemoteConfiguration.serialization.in: Change data type to AtomString
* Source/WebKit/WebProcess/GPU/media/AudioTrackPrivateRemote.cpp:
(WebKit::AudioTrackPrivateRemote::updateConfiguration):
* Source/WebKit/WebProcess/GPU/media/AudioTrackPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/TextTrackPrivateRemote.cpp:
(WebKit::TextTrackPrivateRemote::updateConfiguration):
(WebKit::TextTrackPrivateRemote::addGenericCue):
(WebKit::TextTrackPrivateRemote::updateGenericCue):
(WebKit::TextTrackPrivateRemote::removeGenericCue):
(WebKit::TextTrackPrivateRemote::parseWebVTTFileHeader):
(WebKit::TextTrackPrivateRemote::parseWebVTTCueData):
(WebKit::TextTrackPrivateRemote::parseWebVTTCueDataStruct):
(WebKit::TextTrackPrivateRemote::addDataCue):
(WebKit::TextTrackPrivateRemote::addDataCueWithType):
(WebKit::TextTrackPrivateRemote::updateDataCue):
(WebKit::TextTrackPrivateRemote::removeDataCue):
* Source/WebKit/WebProcess/GPU/media/TextTrackPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/VideoTrackPrivateRemote.cpp:
(WebKit::VideoTrackPrivateRemote::updateConfiguration):
* Source/WebKit/WebProcess/GPU/media/VideoTrackPrivateRemote.h:

Canonical link: https://commits.webkit.org/274956@main
